From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bacteriawa <A3167717663@hotmail.com>
Date: Mon, 4 May 2026 22:31:23 +0800
Subject: [PATCH] Supported Scoreboard And Schedule Command


diff --git a/net/minecraft/commands/Commands.java b/net/minecraft/commands/Commands.java
index 78f1a2c1a92356e800dc0a4865812bef4e1d4f38..6693dac1cbd2e880d4f84949fe9f29e5929cb5e9 100644
--- a/net/minecraft/commands/Commands.java
+++ b/net/minecraft/commands/Commands.java
@@ -239,8 +239,8 @@ public class Commands {
         RideCommand.register(this.dispatcher);
         RotateCommand.register(this.dispatcher);
         SayCommand.register(this.dispatcher);
-        //ScheduleCommand.register(this.dispatcher); // Folia - region threading
-        //ScoreboardCommand.register(this.dispatcher, context); // Folia - region threading
+        if (dev.bacteriawa.mint.config.modules.experiment.ScheduleCommandConfig.enabled) ScheduleCommand.register(this.dispatcher); // Folia - region threading
+        if (dev.bacteriawa.mint.config.modules.experiment.ScoreboardCommandConfig.enabled) ScoreboardCommand.register(this.dispatcher, context); // Folia - region threading
         SeedCommand.register(this.dispatcher, commandSelection != Commands.CommandSelection.INTEGRATED);
         VersionCommand.register(this.dispatcher, commandSelection != Commands.CommandSelection.INTEGRATED);
         SetBlockCommand.register(this.dispatcher, context);
