From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bacteriawa <A3167717663@hotmail.com>
Date: Sun, 10 May 2026 02:26:42 +0800
Subject: [PATCH] PufferFish Dynamic Activation Brain


diff --git a/io/papermc/paper/entity/activation/ActivationRange.java b/io/papermc/paper/entity/activation/ActivationRange.java
index 4cccfb0871fa48a67e6a2f985f475c1b300f718c..8a96250014879401b01ee9dc73a6f0e3061db424 100644
--- a/io/papermc/paper/entity/activation/ActivationRange.java
+++ b/io/papermc/paper/entity/activation/ActivationRange.java
@@ -1,5 +1,7 @@
 package io.papermc.paper.entity.activation;
 
+import dev.bacteriawa.mint.config.modules.optimizations.DynamicActivationBrainConfig;
+
 import net.minecraft.core.BlockPos;
 import net.minecraft.server.MinecraftServer;
 import net.minecraft.world.entity.Entity;
@@ -175,6 +177,24 @@ public final class ActivationRange {
                 }
 
                 ActivationRange.activateEntity(entity, bbByType); // Folia - threaded regions
+                // PufferFish DAB
+                if (DynamicActivationBrainConfig.enabled && entity.getType().dabEnabled()) {
+                    if (!entity.activatedPriorityReset) {
+                        entity.activatedPriorityReset = true;
+                        entity.activatedPriority = DynamicActivationBrainConfig.maximumActivationPrio;
+                    }
+                    net.minecraft.world.phys.Vec3 playerVec = player.position();
+                    net.minecraft.world.phys.Vec3 entityVec = entity.position();
+                    double diffX = playerVec.x - entityVec.x, diffY = playerVec.y - entityVec.y, diffZ = playerVec.z - entityVec.z;
+                    int squaredDistance = (int) (diffX * diffX + diffY * diffY + diffZ * diffZ);
+                    int startDistance = DynamicActivationBrainConfig.startDistance;
+                    entity.activatedPriority = squaredDistance > startDistance * startDistance ?
+                            Math.max(1, Math.min(squaredDistance >> DynamicActivationBrainConfig.activationDistanceMod, entity.activatedPriority)) :
+                            1;
+                } else {
+                    entity.activatedPriority = 1;
+                }
+                // PufferFish DAB
             }
         }
     }
diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index 187f621a4ad5b18dd2e5179eef9f8db73338e19d..0c5813c9e5d12b950bb9969a9257a15544126a19 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -917,6 +917,7 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
             regionizedWorldData // Folia - regionised ticking
                 .forEachTickingEntity( // Folia - regionised ticking
                     entity -> {
+                        entity.activatedPriorityReset = false; // PufferFish DAB
                         if (!entity.isRemoved()) {
                             if (!tickRateManager.isEntityFrozen(entity)) {
                                 profiler.push("checkDespawn");
diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 269dc5466bc9206e2811e1aef5e14412ceb89a7c..354254bc4d058a92f8c2c1076c68c479ea08b167 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -388,6 +388,8 @@ public abstract class Entity
     public long activatedTick = Integer.MIN_VALUE;
     public boolean isTemporarilyActive;
     public long activatedImmunityTick = Integer.MIN_VALUE;
+    public boolean activatedPriorityReset = false; // PufferFish DAB
+    public int activatedPriority = dev.bacteriawa.mint.config.modules.optimizations.DynamicActivationBrainConfig.maximumActivationPrio; // PufferFish DAB
 
     public void inactiveTick() {
     }
diff --git a/net/minecraft/world/entity/EntityType.java b/net/minecraft/world/entity/EntityType.java
index 6efaa78faca15c7a48bdffb24480d1bac29f3ba8..c2414d2357cd747962fc085bea3c58a9624fa5c8 100644
--- a/net/minecraft/world/entity/EntityType.java
+++ b/net/minecraft/world/entity/EntityType.java
@@ -73,6 +73,11 @@ public class EntityType<T extends Entity> implements EntityTypeTest<Entity, T>,
     private final float spawnDimensionsScale;
     private final FeatureFlagSet requiredFeatures;
     private final boolean allowedInPeaceful;
+    private final boolean dabEnabled; // PufferFish DAB
+
+    public final boolean dabEnabled() {
+        return this.dabEnabled;
+    } // PufferFish DAB
 
     public static Identifier getKey(final EntityType<?> type) {
         return BuiltInRegistries.ENTITY_TYPE.getKey(type);
@@ -100,7 +105,7 @@ public class EntityType<T extends Entity> implements EntityTypeTest<Entity, T>,
         final String descriptionId,
         final Optional<ResourceKey<LootTable>> lootTable,
         final FeatureFlagSet requiredFeatures,
-        final boolean allowedInPeaceful, String id // Folia - profiler
+        final boolean allowedInPeaceful, final boolean dabEnabled, String id // Folia - profiler // PufferFish DAB
     ) {
         // Folia start - profiler
         this.tickTimerId = ca.spottedleaf.leafprofiler.LProfilerRegistry.GLOBAL_REGISTRY.getOrCreateTimer("Entity Tick: " + id);
@@ -123,6 +128,7 @@ public class EntityType<T extends Entity> implements EntityTypeTest<Entity, T>,
         this.lootTable = lootTable;
         this.requiredFeatures = requiredFeatures;
         this.allowedInPeaceful = allowedInPeaceful;
+        this.dabEnabled = dabEnabled; // PufferFish DAB
     }
 
     public @Nullable T spawn(
@@ -602,6 +608,7 @@ public class EntityType<T extends Entity> implements EntityTypeTest<Entity, T>,
         );
         private final DependantName<EntityType<?>, String> descriptionId = id -> Util.makeDescriptionId("entity", id.identifier());
         private boolean allowedInPeaceful = true;
+        private boolean dabEnabled = false; // PufferFish DAB
 
         private Builder(final EntityType.EntityFactory<T> factory, final MobCategory category) {
             this.factory = factory;
@@ -720,6 +727,11 @@ public class EntityType<T extends Entity> implements EntityTypeTest<Entity, T>,
             return this;
         }
 
+        public EntityType.Builder<T> dabEnabled() {
+            this.dabEnabled = true;
+            return this;
+        }
+
         public EntityType<T> build(final ResourceKey<EntityType<?>> name) {
             if (this.serialize) {
                 Util.fetchChoiceType(References.ENTITY_TREE, name.identifier().toString());
@@ -740,7 +752,7 @@ public class EntityType<T extends Entity> implements EntityTypeTest<Entity, T>,
                 this.descriptionId.get(name),
                 this.lootTable.get(name),
                 this.requiredFeatures,
-                this.allowedInPeaceful, name.toString() // Folia - profiler
+                this.allowedInPeaceful, this.dabEnabled, name.toString() // Folia - profiler // PufferFish DAB
             );
         }
     }
diff --git a/net/minecraft/world/entity/EntityTypes.java b/net/minecraft/world/entity/EntityTypes.java
index 95b2af33117400cbfa7c62ddbf46e44ab2b49360..9c2e203dd9b420edc6b260c559c0b96dd9c95bc5 100644
--- a/net/minecraft/world/entity/EntityTypes.java
+++ b/net/minecraft/world/entity/EntityTypes.java
@@ -144,6 +144,49 @@ import net.minecraft.world.item.Items;
 import net.minecraft.world.phys.Vec3;
 
 public class EntityTypes {
+    // PufferFish start - DAB
+    private static final Set<ResourceKey<EntityType<?>>> DAB_ENABLED_TYPES = Set.of(
+        EntityTypeIds.ALLAY,
+        EntityTypeIds.ARMADILLO,
+        EntityTypeIds.AXOLOTL,
+        EntityTypeIds.BEE,
+        EntityTypeIds.BREEZE,
+        EntityTypeIds.CAMEL,
+        EntityTypeIds.CAT,
+        EntityTypeIds.CHICKEN,
+        EntityTypeIds.COW,
+        EntityTypeIds.DOLPHIN,
+        EntityTypeIds.DONKEY,
+        EntityTypeIds.FOX,
+        EntityTypeIds.FROG,
+        EntityTypeIds.GOAT,
+        EntityTypeIds.HOGLIN,
+        EntityTypeIds.HORSE,
+        EntityTypeIds.LLAMA,
+        EntityTypeIds.MOOSHROOM,
+        EntityTypeIds.MULE,
+        EntityTypeIds.OCELOT,
+        EntityTypeIds.PANDA,
+        EntityTypeIds.PARROT,
+        EntityTypeIds.PIG,
+        EntityTypeIds.PIGLIN,
+        EntityTypeIds.PIGLIN_BRUTE,
+        EntityTypeIds.POLAR_BEAR,
+        EntityTypeIds.PUFFERFISH,
+        EntityTypeIds.RABBIT,
+        EntityTypeIds.SHEEP,
+        EntityTypeIds.SNIFFER,
+        EntityTypeIds.STRIDER,
+        EntityTypeIds.TADPOLE,
+        EntityTypeIds.TRADER_LLAMA,
+        EntityTypeIds.TURTLE,
+        EntityTypeIds.VILLAGER,
+        EntityTypeIds.WANDERING_TRADER,
+        EntityTypeIds.WARDEN,
+        EntityTypeIds.WOLF,
+        EntityTypeIds.ZOGLIN
+    );
+    // PufferFish end - DAB
     private static final float MAGIC_HORSE_WIDTH = 1.3964844F;
     private static final int DISPLAY_TRACKING_RANGE = 10;
     public static final EntityType<Boat> ACACIA_BOAT = register(
@@ -1200,6 +1243,9 @@ public class EntityTypes {
     }
 
     private static <T extends Entity> EntityType<T> register(final ResourceKey<EntityType<?>> id, final EntityType.Builder<T> builder) {
+        if (DAB_ENABLED_TYPES.contains(id)) {
+            builder.dabEnabled();
+        }
         return Registry.register(BuiltInRegistries.ENTITY_TYPE, id, builder.build(id));
     }
 }
diff --git a/net/minecraft/world/entity/Mob.java b/net/minecraft/world/entity/Mob.java
index 6480de03aa53346a73261428eb94fcea3c5981e3..991fea1cbcbb69b1a99802176182bce140222eb2 100644
--- a/net/minecraft/world/entity/Mob.java
+++ b/net/minecraft/world/entity/Mob.java
@@ -220,10 +220,10 @@ public abstract class Mob extends LivingEntity implements Targeting, EquipmentUs
     public void inactiveTick() {
         super.inactiveTick();
         if (!this.aware) return; // Paper - Do not tick AI for inactive unaware mobs
-        if (this.goalSelector.inactiveTick()) {
+        if (this.goalSelector.inactiveTick(this.activatedPriority, false)) { // PufferFish DAB
             this.goalSelector.tick();
         }
-        if (this.targetSelector.inactiveTick()) {
+        if (this.targetSelector.inactiveTick(this.activatedPriority, false)) { // PufferFish DAB
             this.targetSelector.tick();
         }
     }
@@ -842,16 +842,20 @@ public abstract class Mob extends LivingEntity implements Targeting, EquipmentUs
         profiler.pop();
         if (i % 2 != 0 && this.tickCount > 1) {
             profiler.push("targetSelector");
+            if (this.targetSelector.inactiveTick(this.activatedPriority, true)) // PufferFish DAB
             this.targetSelector.tickRunningGoals(false);
             profiler.pop();
             profiler.push("goalSelector");
+            if (this.goalSelector.inactiveTick(this.activatedPriority, true)) // PufferFish DAB
             this.goalSelector.tickRunningGoals(false);
             profiler.pop();
         } else {
             profiler.push("targetSelector");
+            if (this.targetSelector.inactiveTick(this.activatedPriority, true)) // PufferFish DAB
             this.targetSelector.tick();
             profiler.pop();
             profiler.push("goalSelector");
+            if (this.goalSelector.inactiveTick(this.activatedPriority, true)) // PufferFish DAB
             this.goalSelector.tick();
             profiler.pop();
         }
diff --git a/net/minecraft/world/entity/ai/behavior/VillagerPanicTrigger.java b/net/minecraft/world/entity/ai/behavior/VillagerPanicTrigger.java
index bb037ea27688a06bb0d8141ef25d2543d34aba80..f1460b1c21c9ba4f3206b6a02e736ce813b854ed 100644
--- a/net/minecraft/world/entity/ai/behavior/VillagerPanicTrigger.java
+++ b/net/minecraft/world/entity/ai/behavior/VillagerPanicTrigger.java
@@ -36,9 +36,13 @@ public class VillagerPanicTrigger extends Behavior<Villager> {
 
     @Override
     protected void tick(final ServerLevel level, final Villager body, final long timestamp) {
-        if (timestamp % 100L == 0L) {
+        // PufferFish DAB - start
+        if (body.nextGolemPanic <= 0) {
             body.spawnGolemIfNeeded(level, timestamp, 3);
+            body.nextGolemPanic = 100;
         }
+        body.nextGolemPanic -= body.activatedPriority;
+        // PufferFish DAB - end
     }
 
     public static boolean hasHostile(final LivingEntity myBody) {
diff --git a/net/minecraft/world/entity/ai/goal/GoalSelector.java b/net/minecraft/world/entity/ai/goal/GoalSelector.java
index 0439543fbb25ddb974550fe9dbdaa479e330937d..7d89fc4c24beda23acdda1e8281fc30c7084778e 100644
--- a/net/minecraft/world/entity/ai/goal/GoalSelector.java
+++ b/net/minecraft/world/entity/ai/goal/GoalSelector.java
@@ -42,10 +42,15 @@ public class GoalSelector {
     }
 
     // Paper start - EAR 2
-    public boolean inactiveTick() {
+    // Pufferfish start - DAB
+    public boolean inactiveTick(int activatedPriority, boolean active) {
         this.curRate++;
-        return this.curRate % 3 == 0; // TODO newGoalRate was already unused in 1.20.4, check if this is correct
+        if (!dev.bacteriawa.mint.config.modules.optimizations.DynamicActivationBrainConfig.enabled || !active) {
+            return this.curRate % 20 == 0;
+        }
+        return this.curRate % activatedPriority == 0;
     }
+    // Pufferfish end - DAB
 
     public boolean hasTasks() {
         for (final WrappedGoal task : this.availableGoals) {
diff --git a/net/minecraft/world/entity/animal/allay/Allay.java b/net/minecraft/world/entity/animal/allay/Allay.java
index 69d8a10f524dd4db683a00f194e8d1a26c996f43..026b56f22d1e5ae8d01a902d16e02600b5646482 100644
--- a/net/minecraft/world/entity/animal/allay/Allay.java
+++ b/net/minecraft/world/entity/animal/allay/Allay.java
@@ -101,6 +101,7 @@ public class Allay extends PathfinderMob implements InventoryCarrier, VibrationS
     private float spinningAnimationTicks;
     private float spinningAnimationTicks0;
     public boolean forceDancing = false; // CraftBukkit
+    private int behaviorTick = 0; // PufferFish DAB
 
     public Allay(final EntityType<? extends Allay> type, final Level level) {
         super(type, level);
@@ -210,6 +211,7 @@ public class Allay extends PathfinderMob implements InventoryCarrier, VibrationS
     protected void customServerAiStep(final ServerLevel level) {
         ProfilerFiller profiler = Profiler.get();
         profiler.push("allayBrain");
+        if (this.behaviorTick++ % this.activatedPriority == 0) // PufferFish DAB
         this.getBrain().tick(level, this);
         profiler.pop();
         profiler.push("allayActivityUpdate");
diff --git a/net/minecraft/world/entity/animal/armadillo/Armadillo.java b/net/minecraft/world/entity/animal/armadillo/Armadillo.java
index c8c50683597a2314a3c0e72888d5d75dc73fb94a..437ad8e9dfbea16690f48cfcd571897b796ae139 100644
--- a/net/minecraft/world/entity/animal/armadillo/Armadillo.java
+++ b/net/minecraft/world/entity/animal/armadillo/Armadillo.java
@@ -76,6 +76,7 @@ public class Armadillo extends Animal {
     public final AnimationState rollUpAnimationState = new AnimationState();
     public final AnimationState peekAnimationState = new AnimationState();
     private int scuteTime;
+    private int behaviorTick = 0; // PufferFish DAB
     private boolean peekReceivedClient = false;
 
     public Armadillo(final EntityType<? extends Animal> type, final Level level) {
@@ -142,6 +143,7 @@ public class Armadillo extends Animal {
     protected void customServerAiStep(final ServerLevel level) {
         ProfilerFiller profiler = Profiler.get();
         profiler.push("armadilloBrain");
+        if (this.behaviorTick++ % this.activatedPriority == 0) // PufferFish DAB
         this.getBrain().tick(level, this);
         profiler.pop();
         profiler.push("armadilloActivityUpdate");
diff --git a/net/minecraft/world/entity/animal/axolotl/Axolotl.java b/net/minecraft/world/entity/animal/axolotl/Axolotl.java
index fa774ac8f3eea281c9b684c76dd12d4e66d3ee24..535bcd5927aeecf695a3a90526ed0cf6035ab686 100644
--- a/net/minecraft/world/entity/animal/axolotl/Axolotl.java
+++ b/net/minecraft/world/entity/animal/axolotl/Axolotl.java
@@ -76,6 +76,7 @@ import org.jspecify.annotations.Nullable;
 public class Axolotl extends Animal implements Bucketable {
     public static final int TOTAL_PLAYDEAD_TIME = 200;
     private static final int POSE_ANIMATION_TICKS = 10;
+    private int behaviorTick = 0; // PufferFish DAB
     private static final Brain.Provider<Axolotl> BRAIN_PROVIDER = Brain.provider(
         List.of(SensorType.NEAREST_LIVING_ENTITIES, SensorType.NEAREST_ADULT, SensorType.HURT_BY, SensorType.AXOLOTL_ATTACKABLES, SensorType.FOOD_TEMPTATIONS),
         var0 -> AxolotlAi.getActivities()
@@ -370,6 +371,7 @@ public class Axolotl extends Animal implements Bucketable {
     protected void customServerAiStep(final ServerLevel level) {
         ProfilerFiller profiler = Profiler.get();
         profiler.push("axolotlBrain");
+        if (this.behaviorTick++ % this.activatedPriority == 0) // PufferFish DAB
         this.getBrain().tick(level, this);
         profiler.pop();
         profiler.push("axolotlActivityUpdate");
diff --git a/net/minecraft/world/entity/animal/camel/Camel.java b/net/minecraft/world/entity/animal/camel/Camel.java
index 57127335a3c68fbf9ac3d64a8504c6e271a5d532..b0d072961d3b7c5faef2e3c10451b8128d7ac448 100644
--- a/net/minecraft/world/entity/animal/camel/Camel.java
+++ b/net/minecraft/world/entity/animal/camel/Camel.java
@@ -90,6 +90,7 @@ public class Camel extends AbstractHorse {
         .withEyeHeight(0.845F);
     private int dashCooldown = 0;
     private int idleAnimationTimeout = 0;
+    private int behaviorTick = 0; // PufferFish DAB
 
     public Camel(final EntityType<? extends Camel> type, final Level level) {
         super(type, level);
@@ -172,8 +173,8 @@ public class Camel extends AbstractHorse {
     protected void customServerAiStep(final ServerLevel level) {
         ProfilerFiller profiler = Profiler.get();
         profiler.push("camelBrain");
-        Brain<Camel> brain = this.getBrain();
-        brain.tick(level, this);
+        if (this.behaviorTick++ % this.activatedPriority == 0) // PufferFish DAB
+        this.getBrain().tick(level, this);
         profiler.pop();
         profiler.push("camelActivityUpdate");
         CamelAi.updateActivity(this);
diff --git a/net/minecraft/world/entity/animal/frog/Frog.java b/net/minecraft/world/entity/animal/frog/Frog.java
index d7d8216524e8e4c66c9e5a8dd4e682f46f335ca1..1738392f00355da92e54ec37ac5ae1218b1c0184 100644
--- a/net/minecraft/world/entity/animal/frog/Frog.java
+++ b/net/minecraft/world/entity/animal/frog/Frog.java
@@ -83,6 +83,7 @@ public class Frog extends Animal {
     public final AnimationState croakAnimationState = new AnimationState();
     public final AnimationState tongueAnimationState = new AnimationState();
     public final AnimationState swimIdleAnimationState = new AnimationState();
+    private int behaviorTick = 0; // PufferFish DAB
 
     public Frog(final EntityType<? extends Animal> type, final Level level) {
         super(type, level);
@@ -177,6 +178,7 @@ public class Frog extends Animal {
     protected void customServerAiStep(final ServerLevel level) {
         ProfilerFiller profiler = Profiler.get();
         profiler.push("frogBrain");
+        if (this.behaviorTick++ % this.activatedPriority == 0) // PufferFish DAB
         this.getBrain().tick(level, this);
         profiler.pop();
         profiler.push("frogActivityUpdate");
diff --git a/net/minecraft/world/entity/animal/frog/Tadpole.java b/net/minecraft/world/entity/animal/frog/Tadpole.java
index 296cae1d5f5a3a054949ee03e2a2552bfeae4682..5107401d80b5b45867a2b13d91331d3e670390e2 100644
--- a/net/minecraft/world/entity/animal/frog/Tadpole.java
+++ b/net/minecraft/world/entity/animal/frog/Tadpole.java
@@ -51,6 +51,7 @@ public class Tadpole extends AbstractFish {
     public static final float HITBOX_HEIGHT = 0.3F;
     public int age = 0;
     protected int ageLockParticleTimer = 0;
+    private int behaviorTick = 0; // PufferFish DAB
     private static final Brain.Provider<Tadpole> BRAIN_PROVIDER = Brain.provider(
         List.of(SensorType.NEAREST_LIVING_ENTITIES, SensorType.NEAREST_PLAYERS, SensorType.HURT_BY, SensorType.FROG_TEMPTATIONS),
         var0 -> TadpoleAi.getActivities()
@@ -86,6 +87,7 @@ public class Tadpole extends AbstractFish {
     protected void customServerAiStep(final ServerLevel level) {
         ProfilerFiller profiler = Profiler.get();
         profiler.push("tadpoleBrain");
+        if (this.behaviorTick++ % this.activatedPriority == 0) // PufferFish DAB
         this.getBrain().tick(level, this);
         profiler.pop();
         profiler.push("tadpoleActivityUpdate");
diff --git a/net/minecraft/world/entity/animal/goat/Goat.java b/net/minecraft/world/entity/animal/goat/Goat.java
index d6ad4b67704b5c8b32d064c94890710d8d2f1e57..8cb0c28f2e8f75410ea5c133bdcf23c298339bca 100644
--- a/net/minecraft/world/entity/animal/goat/Goat.java
+++ b/net/minecraft/world/entity/animal/goat/Goat.java
@@ -87,6 +87,7 @@ public class Goat extends Animal {
     private static final boolean DEFAULT_HAS_RIGHT_HORN = true;
     private boolean isLoweringHead;
     private int lowerHeadTick;
+    private int behaviorTick = 0; // PufferFish DAB
 
     public Goat(final EntityType<? extends Goat> type, final Level level) {
         super(type, level);
@@ -176,6 +177,7 @@ public class Goat extends Animal {
     protected void customServerAiStep(final ServerLevel level) {
         ProfilerFiller profiler = Profiler.get();
         profiler.push("goatBrain");
+        if (this.behaviorTick++ % this.activatedPriority == 0) // PufferFish DAB
         this.getBrain().tick(level, this);
         profiler.pop();
         profiler.push("goatActivityUpdate");
diff --git a/net/minecraft/world/entity/animal/sniffer/Sniffer.java b/net/minecraft/world/entity/animal/sniffer/Sniffer.java
index d5394ae7ce56555ad21aeafb2d78c291c62e2988..6c4662c642b4c14b897c2bf0d11fdf73e870f9cc 100644
--- a/net/minecraft/world/entity/animal/sniffer/Sniffer.java
+++ b/net/minecraft/world/entity/animal/sniffer/Sniffer.java
@@ -82,6 +82,7 @@ public class Sniffer extends Animal {
     public final AnimationState sniffingAnimationState = new AnimationState();
     public final AnimationState diggingAnimationState = new AnimationState();
     public final AnimationState risingAnimationState = new AnimationState();
+    private int behaviorTick = 0; // PufferFish DAB
 
     public static AttributeSupplier.Builder createAttributes() {
         return Animal.createAnimalAttributes().add(Attributes.MOVEMENT_SPEED, 0.1F).add(Attributes.MAX_HEALTH, 14.0);
@@ -468,6 +469,7 @@ public class Sniffer extends Animal {
     protected void customServerAiStep(final ServerLevel level) {
         ProfilerFiller profiler = Profiler.get();
         profiler.push("snifferBrain");
+        if (this.behaviorTick++ % this.activatedPriority == 0) // PufferFish DAB
         this.getBrain().tick(level, this);
         profiler.popPush("snifferActivityUpdate");
         SnifferAi.updateActivity(this);
diff --git a/net/minecraft/world/entity/monster/breeze/Breeze.java b/net/minecraft/world/entity/monster/breeze/Breeze.java
index eaa7eb69bdc819228bbab05b053844f1578294ba..f259d2f042c51201454f5076609bab3bcbbfa3cf 100644
--- a/net/minecraft/world/entity/monster/breeze/Breeze.java
+++ b/net/minecraft/world/entity/monster/breeze/Breeze.java
@@ -59,6 +59,7 @@ public class Breeze extends Monster {
     public final AnimationState inhale = new AnimationState();
     private int jumpTrailStartedTick = 0;
     private int soundTick = 0;
+    private int behaviorTick = 0; // PufferFish DAB
     private static final ProjectileDeflection PROJECTILE_DEFLECTION = (projectile, entity, random) -> {
         entity.level().playSound(null, entity, SoundEvents.BREEZE_DEFLECT, entity.getSoundSource(), 1.0F, 1.0F);
         ProjectileDeflection.REVERSE.deflect(projectile, entity, random);
@@ -234,6 +235,7 @@ public class Breeze extends Monster {
     protected void customServerAiStep(final ServerLevel level) {
         ProfilerFiller profiler = Profiler.get();
         profiler.push("breezeBrain");
+        if (this.behaviorTick++ % this.activatedPriority == 0) // PufferFish DAB
         this.getBrain().tick(level, this);
         profiler.popPush("breezeActivityUpdate");
         BreezeAi.updateActivity(this);
diff --git a/net/minecraft/world/entity/monster/hoglin/Hoglin.java b/net/minecraft/world/entity/monster/hoglin/Hoglin.java
index 045b8320ef9d63d7b81a680ad12f1afa94a76c7d..84f5304aae404ee5d6fafe9fe1e44f9d6f82bc2b 100644
--- a/net/minecraft/world/entity/monster/hoglin/Hoglin.java
+++ b/net/minecraft/world/entity/monster/hoglin/Hoglin.java
@@ -71,6 +71,7 @@ public class Hoglin extends Animal implements Enemy, HoglinBase {
     private int attackAnimationRemainingTicks;
     public int timeInOverworld = 0;
     public boolean cannotBeHunted = false;
+    private int behaviorTick = 0; // PufferFish DAB
     private static final Brain.Provider<Hoglin> BRAIN_PROVIDER = Brain.provider(
         List.of(SensorType.NEAREST_LIVING_ENTITIES, SensorType.NEAREST_PLAYERS, SensorType.NEAREST_ADULT, SensorType.HOGLIN_SPECIFIC_SENSOR),
         var0 -> HoglinAi.getActivities()
@@ -144,6 +145,7 @@ public class Hoglin extends Animal implements Enemy, HoglinBase {
     protected void customServerAiStep(final ServerLevel level) {
         ProfilerFiller profiler = Profiler.get();
         profiler.push("hoglinBrain");
+        if (this.behaviorTick++ % this.activatedPriority == 0) // PufferFish DAB
         this.getBrain().tick(level, this);
         profiler.pop();
         HoglinAi.updateActivity(this);
diff --git a/net/minecraft/world/entity/monster/piglin/Piglin.java b/net/minecraft/world/entity/monster/piglin/Piglin.java
index 77638e1b9b2ab1165d11c1ac8ea09255b04544f6..605333c08c7654a9c152533ea9bc002fdc915f53 100644
--- a/net/minecraft/world/entity/monster/piglin/Piglin.java
+++ b/net/minecraft/world/entity/monster/piglin/Piglin.java
@@ -85,6 +85,7 @@ public class Piglin extends AbstractPiglin implements CrossbowAttackMob, Invento
     private static final int INVENTORY_SIZE = 8;
     private final SimpleContainer inventory = new SimpleContainer(8);
     public boolean cannotHunt = false;
+    private int behaviorTick = 0; // PufferFish DAB
     private static final Brain.Provider<Piglin> BRAIN_PROVIDER = Brain.provider(
         List.of(
             MemoryModuleType.UNIVERSAL_ANGER,
@@ -284,6 +285,7 @@ public class Piglin extends AbstractPiglin implements CrossbowAttackMob, Invento
     protected void customServerAiStep(final ServerLevel level) {
         ProfilerFiller profiler = Profiler.get();
         profiler.push("piglinBrain");
+        if (this.behaviorTick++ % this.activatedPriority == 0) // PufferFish DAB
         this.getBrain().tick(level, this);
         profiler.pop();
         PiglinAi.updateActivity(this);
diff --git a/net/minecraft/world/entity/monster/piglin/PiglinBrute.java b/net/minecraft/world/entity/monster/piglin/PiglinBrute.java
index 8bba57ee432ab41b47b6dd89a21bf17c4b462d1f..b0dfe34b9381754461a8842a6d57ef18349bc72e 100644
--- a/net/minecraft/world/entity/monster/piglin/PiglinBrute.java
+++ b/net/minecraft/world/entity/monster/piglin/PiglinBrute.java
@@ -49,6 +49,7 @@ public class PiglinBrute extends AbstractPiglin {
         super(type, level);
         this.xpReward = 20;
     }
+    private int behaviorTick = 0; // PufferFish DAB
 
     public static AttributeSupplier.Builder createAttributes() {
         return Monster.createMonsterAttributes()
@@ -96,6 +97,7 @@ public class PiglinBrute extends AbstractPiglin {
     protected void customServerAiStep(final ServerLevel level) {
         ProfilerFiller profiler = Profiler.get();
         profiler.push("piglinBruteBrain");
+        if (this.behaviorTick++ % this.activatedPriority == 0) // PufferFish DAB
         this.getBrain().tick(level, this);
         profiler.pop();
         PiglinBruteAi.updateActivity(this);
diff --git a/net/minecraft/world/entity/monster/warden/Warden.java b/net/minecraft/world/entity/monster/warden/Warden.java
index 4552a36c3cd9b5c0f34cd44d376d6d1092ee110a..5218e89242488f79ce15c2b672a136f7d4b57414 100644
--- a/net/minecraft/world/entity/monster/warden/Warden.java
+++ b/net/minecraft/world/entity/monster/warden/Warden.java
@@ -119,6 +119,7 @@ public class Warden extends Monster implements VibrationSystem {
     private final VibrationSystem.User vibrationUser;
     private VibrationSystem.Data vibrationData;
     private AngerManagement angerManagement = new AngerManagement(this::canTargetEntity, Collections.emptyList());
+    private int behaviorTick = 0; // PufferFish DAB
 
     public Warden(final EntityType<? extends Monster> type, final Level level) {
         super(type, level);
@@ -289,6 +290,7 @@ public class Warden extends Monster implements VibrationSystem {
     protected void customServerAiStep(final ServerLevel level) {
         ProfilerFiller profiler = Profiler.get();
         profiler.push("wardenBrain");
+        if (this.behaviorTick++ % this.activatedPriority == 0) // PufferFish DAB
         this.getBrain().tick(level, this);
         profiler.pop();
         super.customServerAiStep(level);
diff --git a/net/minecraft/world/entity/npc/villager/Villager.java b/net/minecraft/world/entity/npc/villager/Villager.java
index 660b5311ec7bdf37fcb75ead23d59ec4b66f8380..6ae3c2f650b3a5f0ee0634dd4bf458088cc160e9 100644
--- a/net/minecraft/world/entity/npc/villager/Villager.java
+++ b/net/minecraft/world/entity/npc/villager/Villager.java
@@ -127,6 +127,8 @@ public class Villager extends AbstractVillager implements VillagerDataHolder, Re
     public int numberOfRestocksToday = 0;
     private long lastRestockCheckDay;
     private boolean assignProfessionWhenSpawned = false;
+    private int behaviorTick = 0; // PufferFish DAB
+    public int nextGolemPanic = 0; // Pufferfish DAB
     private static final Brain.Provider<Villager> BRAIN_PROVIDER = Brain.provider(
         List.of(
             SensorType.NEAREST_LIVING_ENTITIES,
@@ -264,7 +266,8 @@ public class Villager extends AbstractVillager implements VillagerDataHolder, Re
         // Paper end - EAR 2
         ProfilerFiller profiler = Profiler.get();
         profiler.push("villagerBrain");
-        if (!inactive) this.getBrain().tick(level, this); // Paper - EAR 2
+        if (!inactive && this.behaviorTick++ % this.activatedPriority == 0) // PufferFish DAB
+        this.getBrain().tick(level, this); // Paper - EAR 2
         profiler.pop();
         if (this.assignProfessionWhenSpawned) {
             this.assignProfessionWhenSpawned = false;
