From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bacteriawa <A3167717663@hotmail.com>
Date: Tue, 5 May 2026 01:15:22 +0800
Subject: [PATCH] Add Networkbar


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index ca1f315010c526cb2ddbb1c4708ba8f99c6e4d3a..c48a64014b47b46ef4baf4d0f72395a8ddedd2e6 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -2207,6 +2207,9 @@ public class CraftPlayer extends CraftHumanEntity implements Player, PluginMessa
             //Luminol start - Regionbar
             getHandle().isRegionBarVisible = data.getBooleanOr("regionbarVisible", false);
             //Luminol end
+            // Mint start - Networkbar
+            getHandle().isNetworkBarVisible = data.getBooleanOr("networkbarVisible", false);
+            // Mint end - Networkbar
         });
     }
 
@@ -2233,6 +2236,9 @@ public class CraftPlayer extends CraftHumanEntity implements Player, PluginMessa
         //Luminol start - Regionbar
         data.putBoolean("regionbarVisible", handle.isRegionBarVisible);
         //Luminol end
+        // Mint start - Networkbar
+        data.putBoolean("networkbarVisible", handle.isNetworkBarVisible);
+        // Mint end - Networkbar
 
         // Paper start - persist for use in offline save data
         ValueOutput paper = output.child("Paper");
